service/evaluation/evaluation.go

```diff
@@ -0,0 +1,845 @@
+	if !all {
+		query = append(query, "cloud_service_id IN ?")
+		args = append(args, allowed)
+	}
+
+	// If we want to have it grouped by resource ID, we need to do a raw query
+	if req.GetLatestByResourceId() {
+		// In the raw SQL, we need to build the whole WHERE statement
+		var where string
+		var p = ""
+
+		if len(query) > 0 {
+			where = "WHERE " + strings.Join(query, " AND ")
+		}
+
+		if len(partition) > 0 {
+			p = ", " + strings.Join(partition, ",")
+		}
+
+		// Execute the raw SQL statement
+		err = s.storage.Raw(&res.Results,
+			fmt.Sprintf(`WITH sorted_results AS (
+				SELECT *, ROW_NUMBER() OVER (PARTITION BY resource_id %s ORDER BY timestamp DESC) AS row_number
+				FROM evaluation_results
+				%s
+		  	)
+		  	SELECT * FROM sorted_results WHERE row_number = 1;`, p, where), args...)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "database error: %v", err)
+		}
+	} else {
+		// join query with AND and prepend the query
+		args = append([]any{strings.Join(query, " AND ")}, args...)
+
+		// Paginate the results according to the request
+		res.Results, res.NextPageToken, err = service.PaginateStorage[*evaluation.EvaluationResult](req, s.storage, service.DefaultPaginationOpts, args...)
+		if err != nil {
+			err = fmt.Errorf("could not paginate evaluation results: %v", err)
+			log.Error(err)
+			return nil, status.Errorf(codes.Internal, "could not paginate results: %v", err)
+		}
+	}
+
+	return
+}
+
+// addJobToScheduler adds a job for the given control to the scheduler and sets the scheduler interval to the given interval
+func (s *Service) addJobToScheduler(c *orchestrator.Control, toe *orchestrator.TargetOfEvaluation, parentSchedulerTag string, interval int) (err error) {
+	// Check inputs and log error
+	if c == nil {
+		err = errors.New("control is invalid")
+	}
+	if toe == nil {
+		err = errors.New("target of evaluation is invalid")
+	}
+	if interval == 0 {
+		err = errors.New("interval is invalid")
+	}
+	if err != nil {
+		log.Error(err)
+		return status.Errorf(codes.Internal, "evaluation cannot be scheduled: %v", err)
+	}
+
+	// schedulerTag is the tag for the given control
+	schedulerTag := getSchedulerTag(toe.GetCloudServiceId(), c.GetId())
+
+	// Regarding the control level the specific method is called every X minutes based on the given interval. We have to decide if a sub-control is started individually or a parent control that has to wait for the results of the sub-controls.
+	// If a parent control with its sub-controls is started, the parentSchedulerTag is empty and the ParentControlId is not set.
+	// If a sub-control is started individually the parentSchedulerTag is empty and ParentControlId is set.
+	if parentSchedulerTag == "" && c.ParentControlId == nil { // parent control
+		_, err = s.scheduler.
+			Every(interval).
+			Minute().
+			Tag(schedulerTag).
```

Comment:
Here I would add an additional tag about the ToE so we can quickly stop the evaluation again without knowing all the controls

Reply:
I think both approaches work, so what ever you prefer.